### DISCUSSION TITLE
[Feature] Pet detection

### DISCUSSION THREAD

**Discussion by alternativesurfer** [immich-app/immich#7151] [Feature Request]
### The feature

I notice that dogs get picked up with the "people" facial recognition but cats don't.
Any thoughts on how we could add that?

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by arhills** [comment]
One of my cats got detected as a person. The other 3 did not.  None of my dogs got detected. This might not help at all :)

**Comment by aviv926** [comment]
There are few models trained for such an operation, do you know a model good enough for such a purpose?
The model we currently use for facial recognition does not support animal recognition.

As a temporary alternative: you can describe your pet in the search, I believe that the CLIP model on which the search is based will be able to find your pet with a fairly high accuracy.

**Reply by alternativesurfer** [reply]
True, CLIP does appear to search properly.

**Reply by rodrigopolo** [reply]
Here is one:
https://github.com/MarQuisCheshire/Pets-Face-Recognition

**Reply by Chuckame** [reply]
How to use it in immich? Wouldn't it replace the current face detection so we lose human faces?

**Reply by Roachforlife** [reply]
Would be interested as well. Is it possible to add 2 rounds of detection, one for humans and one for pets?

**Comment by tekjoey** [comment]
I just realized Apple Photos does pet detection, so this is a feature I would like for sure. Immich picked up some of my dogs, but not all of them, and not all photos of dogs. I certainly would like to be able to link photos with a pet. My current implementation is manually adding to albums, which works but isn't great.

(Some musings for the devs) If this feature ever does get implemented, I think there should be a per-user setting to display the pet faces on the same page as people, on a separate (pet only) page, or not at all (currently Apple Photos displays all faces together). Perhaps the database could be updated with a new value of "face_type" with options of "person" or "pet" (maybe specify the type of pet? Maybe the type of pet can be configurable by the user?) Also, when a model does get chosen/trained, please remember that cats and dogs are not the only pets that people have! My wife has horses, so that is a must for us, but lots of others exist as well. (again, these are just some ideas. Take all, some, or none as you see fit)

**Comment by gilgwath** [comment]
Being able to recognise our cats would probably quadruple the spouse approval factor. I think about a third of our photos are of our cats :-)

**Comment by joelis10** [comment]
+1 for wanting pet detection

**Comment by mrshmllow** [comment]
Unironically this is blocking me from switching from google photos

**Reply by datasian** [reply]
Same. I do have all of my photos cataloged onto immich finally but Google Photos helps me find specific pets to a degree. Google Photos does recognition and lets me search by their proper name. That's a killer feature if I'm looking for two specific cats in the same photo like the twins doing cute twin things. I do have unique cats that a general recognition would find. One Siamese, one void, one nebelung and one tuxedo. Distinct colorings helps it. Google Photos will mix up my late Siamese and my current one, and 4 different tuxedo cats that we either babysat or rescued but that's something acceptable for a pet recognition feature.

**Comment by VACInc** [comment]
It's worth noting that it may drive user adoption adding this properly.  Google Photos does this at a very basic level.  If your pet passes and you get another pet (similar breed), you cannot create a new pet and have it recognize it as a new pet.  It continues to tag it as the old pet.  No way to delete said old pet either.

This is what drove me to at least find Immich (at which point I realized there was zero pet recognition).

**Comment by TAT-Hins** [comment]
Hi everyone. This is my first time in the discussion, so sorry if I'm not expressing in the right way.

> It's worth noting that it may drive user adoption adding this properly. Google Photos does this at a very basic level. If your pet passes and you get another pet (similar breed), you cannot create a new pet and have it recognize it as a new pet. It continues to tag it as the old pet. No way to delete said old pet either.
> 
> This is what drove me to at least find Immich (at which point I realized there was zero pet recognition).

I have a cat and two dogs. It's very easy to use smart search by ML clip models to locate the photos of them, so I suppose it should also work with facial recognition, so that I can tag and name them just like people to make it more easily, which seems to be a possible solution for the doubt.

I'm pretty sure that the feature is technically feasible, and I notice that Google Photos has this feature for a long time, but I have no idea how much work it should take to achieve this. It would be awesome if immich can make this happen, and I hope to be a part of the contribution, no matter as a code debugger or just as a user.

**Reply by svonjoi** [reply]
> tag and name them just like people to make it more easily

useless for me cuz i would see the age of pet in desc

**Comment by WizenThorne** [comment]
I'd love to have this feature as well.

**Comment by Chuckame** [comment]
Just putting some random links about Dave detection for pets. It seems to be really hard to find on the internet...!

The most advanced source I've found: https://github.com/GuillaumeMougeot/DogFaceNet

ByteDance seems to provide api for cat and dog recognition (at which price?) : https://docs.byteplus.com/en/docs/effects/docs-pet-face-recognition

As generally people have only a few pets, maybe we could more have a clustering of similar dog than pure face recognition 🤔

**Reply by Chuckame** [reply]
For cats : https://levelup.gitconnected.com/detect-your-cats-face-with-opencv-21893210737e

**Reply by nagyv** [reply]
Related model definitions(?) are available at https://www.kaggle.com/datasets/wutheringwang/dog-face-recognition/data

**Comment by KuramaSyu** [comment]
Where in the code is the current face detection done? If it's not too hard, I would consider making a PR. I don't think that it would be so hard to use 2 AI models for scanning images.

**Reply by aviv926** [reply]
I'm pretty sure it's done here
https://github.com/immich-app/immich/tree/main/machine-learning/app/models/facial_recognition

You might want to consult with @mertalev  before starting work on such a PR, to check if it is something that will be accepted for the project.

**Reply by mertalev** [reply]
Adding it to immich is of course fine, but getting a pet recognition (not detection) model in the first place is more involved. There's some discussion on training a model for this on Discord.

**Reply by nagyv** [reply]
@mertalev As [frigate](https://github.com/blakeblackshear/frigate) already ships with models to detect people, cars, cats, dogs out of the box and is MIT licensed, immich might be able to use the same models.

I think they load the models into their docker images around here: https://github.com/blakeblackshear/frigate/blob/dev/docker/main/Dockerfile#L115

**Reply by bo0tzz** [reply]
AFAIK, Frigate's models are detection models ("Is this a cat"), not recognition models ("Is this cat the same as that other one").

**Comment by pilot51** [comment]
I just discovered Immich today (thanks Google feed) and I'm interested in switching from Google Photos to a self-hosted solution. I have it running and I really like what I see, but it'll be hard for me to commit to a full switch if cat detection isn't working.

My cats are like 90% of all the faces recognized in Google Photos, which automatically adds the photos to an album for both cats shared with my family.

**Reply by aviv926** [reply]
> I just discovered Immich today (thanks Google feed) and I'm interested in switching from Google Photos to a self-hosted solution. I have it running and I really like what I see, but it'll be hard for me to commit to a full switch if cat detection isn't working.
> 
> My cats are like 90% of all the faces recognized in Google Photos, which automatically adds the photos to an album for both cats shared with my family.

You can always use the smart search for now (hint, it really is smart) Check it out in its demo (this is the basic model) Read more here:
https://immich.app/docs/features/smart-search

Also:

https://demo.immich.app

**Reply by pilot51** [reply]
> You can always use the smart search for now (hint, it really is smart) Check it out in its demo (this is the basic model) Read more here: https://immich.app/docs/features/smart-search

That works well for generic searches like "cat". However, what if I only want to find photos of a pet by name?

I would also like to ensure accurate results that show all of the desired photos and no false positives. It's practically impossible to be sure with on-demand results like that.
For example, I just searched for "cat" and it's very accurate at the top of the results but quickly becomes interspersed with other animals and eventually random photos with 0% cat-like objects as you scroll down (it's even worse in the demo with only one cat photo), which tells me there's no certainty cut-off.
That's why I like to use albums that I can curate myself, in this case assisted by ML to add detected faces to the album, which is essential for such high volumes of cat photos.

One thing Google Photos really struggles with is identifying which cat is which (they're both the same color but with distinct features), so I regularly have to correct it by removing the wrong detected cat and selecting the correct one. I don't expect any solution here to be much better in the short term, but if anything comes close to that and lets me easily correct detection errors, I'll be happy. Google hasn't improved at all in the 5 years I've had these cats, so I can see the open source community eventually doing it better.

**Reply by Aevaris** [reply]
I 100% would love a way to automatically identify objects and pets, so take my upvote. That said...

> However, what if I only want to find photos of a pet by name?

> I would also like to ensure accurate results that show all of the desired photos and no false positives.

You can't do it by name, but, if you use context searching, it does a better job than Google Photos does in my experience. 

For example, I have 2 cats, both ragdolls, one is the Sealpoint color and one is a Bluepoint. They are fairly different in color but Google Photos consistently mixes them up. However, in Immich, if I enter a search of 'White cat with a dark grey face', it will find my Bluepoint almost perfectly (certainly better than Google does).

It's not automatic, but you could from there either:
1. Add all results to an album so then there is no risk / effort of re-running the search
2. Add a tag to your photos, and then there is no risk / effort of re-running the search.

Edit: I forgot that you can't context search at all in Google. As a result, I forgot that I gave up on the pet feature because it identified my Bluepoint as my only pet and mixes up between my 2 cats (but also fails to identify my other cat as a pet about 50% of the time entirely). Since I am at the mercy of GP's automation, I can't really use it to find my pets. In this way, I've found Immich to perform better.

**Reply by pilot51** [reply]
> You can't do it by name, but, if you use context searching, it does a better job than Google Photos does in my experience.

That wouldn't work for me. Both of my cats are long haired cream colored flame points and the only way I know to describe the difference is in relative terms. One has longer hair, bigger eyes, and slightly darker nose and ears than the other.

GP is pretty bad at telling the difference, but at least it reliably detects that they're one of my cats and automatically adds them to the shared album I have for both cats. Occasionally it fails to detect and I need to manually add the photo, though I can't really blame it when their face isn't showing and they're basically camouflaged in their cream colored bed.

> 1. Add all results to an album so then there is no risk / effort of re-running the search

That wouldn't work either because, as I previously mentioned, there is no certainty cut-off in the results. As I scroll down, it transitions to cats that don't match the description, then other animals, then endless non-animals. As far as I can tell, adding all context-only search results to an album would effectively add my entire library to the album regardless of the search.

> 2. Add a tag to your photos, and then there is no risk / effort of re-running the search.

How is that any better than manually adding every photo of my cats to the shared album, as is the desired end result?

> I forgot that you can't context search at all in Google.

Yes you can. I just tested and photos that matched the context appeared at the top. I compared against Immich with the exact same search terms, like "mouse" or "eagle" (which I knew I had only a handful of photos) and it did far better. For "eagle", GP found 4 of my 5 photos and Immich missed entirely and gave me a bunch of military museum photos at the top. I scrolled quite a ways and didn't see any eagle photos and very few of my many bird photos. GP still has an issue with many results having nothing to do with the search term, but at least it does have a somewhat reasonable certainty cut-off, for example only giving 9 results for "eagle" and 41 (of 16) for "mouse". Though it does appear to use albums for an assist since I do have an album for pests and the results included some other critters in the album. I haven't bothered recreating my albums in Immich yet because it's such a huge chore and I'm procrastinating.

In my experience, GP wins hands down for object detection and automation.

**Reply by mertalev** [reply]
If you're using the default search model, you can try using a different one with [this](https://github.com/immich-app/immich/discussions/11862) guide. There are much better models.

**Comment by david-askari** [comment]
For the record: The Immich team talked about pet recognizion in a recent [livestream](https://youtu.be/6KtNNjp78xI), starting around 17:24.

TL;DR
- No good enough and free model for this task is available
- No training data available to train a model themselves
- Idea to crowdsource the training data from user uploads

**Reply by Aevaris** [reply]
> Idea to crowdsource the training data from user uploads

I'd help.

**Reply by Roachforlife** [reply]
I have loads of pictures of about 20 pets in my library and would be happy to contribute to any training

**Reply by nagyv** [reply]
I just learned about a large cat face database used for cat-ID training: https://arxiv.org/html/2501.02112v1 and https://www.kaggle.com/datasets/tobiastrein/heellostreetcat-individuals

Similarly, there are the [MegaDescriptor models](https://arxiv.org/html/2311.09118v2) ([HuggingFace](https://huggingface.co/BVRA)) that were created primarily for ecologists for species-independent animal identification in the wild.

Could [DeepSeek-VL2](https://github.com/deepseek-ai/DeepSeek-VL2) be used as the basis for pet-identification training?

On a related note: how can I help? (I used to be a Python programmer before moving to product management. I can still write hopefully good enough quality code if needed. I don't have any serious ML training.)

**Comment by daltimus11** [comment]
I just made the switch to Immich and I am extremely impressed. However, pet detection is something I am missing after leaving Google Photos. A lot of people are passionate about their pets, seeing them as part of the family, I imagine this could be important to many.

Having the ability to manually add "people" could be another solution. #7301

**Reply by Roachforlife** [reply]
> I just made the switch to Immich and I am extremely impressed. However, pet detection is something I am missing after leaving Google Photos. A lot of people are passionate about their pets, seeing them as part of the family, I imagine this could be important to many.
> 
> Having the ability to manually add "people" could be another solution. #7301

I agree. If the ability to manually add a person (or animal in this case) and then manually tag the photo to that person, would be great. At least as an alternative until there is a model that does it all automatically

**Comment by eleaner** [comment]
don’t forget about cats!

**Reply by DraculaVladimir** [reply]
And parrots, rats, monkeys, peguins, and tiny little lady bugs

**Comment by allmazz** [comment]
I occasionally found my pets in People after switching to antelopev2 model
![image](https://github.com/user-attachments/assets/30933b90-30eb-4756-8471-2759d870cea9)

Minimum detection score - 0.6
Maximum recognition distance - 0.5
Minimum recognized faces - 1

Of course with mistakes, but I will try to tweak the settings

**Comment by DraculaVladimir** [comment]
Can you guys please add "Pet face Recognition" to the official roadmap?

**Comment by Lesani** [comment]
I 13.798th this, I have switched to immich despite it not being implemented, but as apple photos does it so well, I would really like to have this. Apples implementation even separates nicely two very similar (tuxedo) cats completely accurately. I have not seen a single mixed up photo there.

it could be an additional detection step, maybe even where a user can define it as an optional feature to be enabled, so not every user runs it, only those with and interested in detecting their pets

**Comment by akshat-khare** [comment]
This feature is very much needed. Pet are family and one of the primary usecase of google photos because its so good at categorising pets too. Lets add that!

**Comment by akshat-khare** [comment]
Does all the target user will have infrastructure lying around to power
deepseek. I'm afraid not. It's better to fine tune the already available
and used models and maybe see with them

SAM from meta can be a good model here imho

**Comment by thornygravy** [comment]
this would be nice for sure

**Comment by Knuppel1983** [comment]
How far are we on this? 75% of my database consists of dog photos. We do all our (hiking) holidays with our dogs, they're like our children. We need this feature, right now it's stopping me from ditching Google Photos.

**Reply by bo0tzz** [reply]
This isn't happening anytime soon.

**Reply by akshat-khare** [reply]
How are you sure

**Reply by rouke-broersma** [reply]
Because they work at Immich? And as explained before there are no good open models for pet detection yet.

**Reply by LukeCarrier** [reply]
If you need it so much, contribute it. If you don't plan to, just add a thumbs up to the issue and subscribe if you want updates.

These entitled +1 and "how far are we(!?) on this" comments do nothing but put maintainers and contributors off their valuable work. Please, just stop.

**Reply by Knuppel1983** [reply]
Hold your horses… We all appreciate the work being done and would love to help (at least i would). We’re just not all developers.

**Reply by mertalev** [reply]
We're lucky to be working on a project with so many exciting ideas and possibilities! Unfortunately, we can only work on a tiny fraction of those ideas at any given time.

We look forward to implementing pet recognition when we can, unless a contributor beats us to it 😉 (if this is you, join our Discord!)

**Reply by LukeCarrier** [reply]
> Hold your horses… We all appreciate the work being done and would love to help (at least i would). We’re just not all developers.

I wasn't born a developer, and neither were the maintainers 😉

If you don't care enough to learn, the best thing you can do to support the maintainers is to avoid cluttering issues with +1 comments that add no value to the discussion and actively put off contributors.

**Reply by Knuppel1983** [reply]
I know how it works, i work in IT. The ones that moan the most get what they want. ;-)

**Reply by bo0tzz** [reply]
The ones that moan the most get their comments minimized as duplicate by annoyed maintainers ;)

**Reply by pilot51** [reply]
I've done IT and currently software development. They're very different beasts when it comes to solving problems.

Typically, IT issues can be solved within a few hours and it's fairly easy to keep up and get some downtime.

In software dev, on the other hand, issues typically take days or weeks but have a very wide range between minutes and years. That means for many if not most active public projects, new issues are created at a higher rate than there is capacity to work through them, so they need to be triaged and many are ignored if they aren't a major bug.

Moaning rarely helps prioritize an issue. It's about severity (including votes), how easy it is to resolve, and especially for volunteer projects, how much the dev is personally interested in working on the issue.

Also, just because I'm a developer with some applicable experience and I want this feature doesn't mean I have the time or passion to contribute. I have a hard enough time getting motivated to work on my own FOSS projects when I have a full-time dev job.

However, I do believe anyone who impatiently moans, especially without adding value to the conversation, would be much better off spending their time learning how to do it themselves. Who knows, maybe they'd kick off a fulfilling career kind of like I did.

**Comment by PPPDUD** [comment]
This would be so cool!

**Reply by PseudoResonance** [reply]
For the sake of everyone's notifications, please do not make comments that don't add anything new... Just upvote the discussion.

I want the feature as much as anyone, but I also don't need daily emails of "+1". We know.

**Comment by nagyv** [comment]
Could AutoGlon's Image-to-Image Semantic Matching be used to train a pet recognising model? https://auto.gluon.ai/stable/tutorials/multimodal/semantic_matching/image2image_matching.html

I'm not an ML researcher. As a result, I only have questions.

Theoretically:

- There were many people here who offered to share their pet images. 
- We could collect pet images (let's say 5-20 images / pet)
- Combine every image with every other image to get a big variation of matches and no-matches
- Train and the model based on these combinations

## Questions for data collection

- How many (individual) pets and how many images / pet would be needed for a meaningful dataset?
- How clean the images should be? (What if one image shows two dogs? Each dog has multiple images, but some of those are "shared".)

**Reply by rtp4jc** [reply]
This is a really unique idea! I hadn't thought of it and I've been trying to think of workarounds to my problems (see my post a few messages below). 

Before I say anything: I hope I'm not telling you too much you already know. Sorry if I am.

**Technically** this might work, but the problem is that it would be computationally expensive compared to existing human identification. What you're describing is after the detection part (we already know the picture has a dog). Most models will create a vector (a list of numbers) that "represent" the given dog. Generating this vector is the expensive part because it has to go through the model. The same dog in two different images should produce a similar vector (the face is the same, but the orientation and lighting might be different). We use this principle to compare the vector to other dog image vectors and see which vectors are closest. This requires a lot of comparisons, but each comparison is cheap.

With the direct image to image comparison, you're having to do an expensive operation on every **pair** of images as opposed to once for each image and then doing a cheap comparison on every pair of vectors.

That being said: would it perform better with the same data? Only testing can tell, but my guess is it would be about the same. 

---

Now about the dataset questions: these are good questions. I have a cleaner answer for your second question so I'll answer that first: the data should be representative of how you'll use it in real life. It should be as messy or as clean of photos as you would have in your image roll. I have, quite literally, professional portraits of my dog in my gallery, but I also have photos of him derping out upside down in a messy room. Both are valid and should be in the dataset. It's also very common to have photos with multiple dogs (technically they will be analyzed by separate models at one point in the identification pipeline, but that's use-case specific) so these should also be in the dataset. 

For your first question, I don't have a good target number off the top of my head, but I can tell you what is NOT enough: the dataset I've been using for dog identification has 1000 identities (unique dogs) with an average of 7 photos per identity. This is the dataset that has resulted in subpar performance in my tests. I would say in order to get closer to a production quality model, we would need something along the lines of 5000-10000 identities (per species!!) with an average of 50-100 photos per identity, but I'm just speculating at that point based on what I've seen. It could be that a truly representative dataset (lots of variety in identity and between photos of the same identity) that is still clean (generally free of human error) might allow for a significantly smaller data requirement.

**Comment by not4smurf** [comment]
I'm just here to add my support for this request.  Well, I actually came for help figuring out how to fix it because it has recognized one of my dogs as "People" but not the other.

**Comment by specter9mm** [comment]
Would also love a pet recognition feature. Also, immich has been absolutely amazing so far!

**Reply by PseudoResonance** [reply]
For the sake of everyone's notifications, please do not make comments that don't add anything new... Just upvote the discussion.

I want the feature as much as anyone, but I also don't need daily emails of "+1". We know.

**Comment by racerxdl** [comment]
Frigate does a sort of good job of detecting pets. Maybe it could be done in two stage: find pets and then classify by similarity. 

FYI: https://docs.frigate.video/plus/ - they have finuted models for frigate+ but the standard models also works great. They're made for way worse images (security cameras). MobileDet and Yolo work great.

Also, this might be able to tag other stuff for contextual search (cars, etc...) and aggregate non-face images.

**Comment by rtp4jc** [comment]
Hey all, I've been working towards an implementation of this that follows patterns used for human face detection and identification. You can find the code here: https://github.com/rtp4jc/immich-animals. 

I'll preface by saying: **it's not good and I've only trained it with dogs so far**. I need to get some metrics added to the repo so people have a baseline, but in terms of actually using it for real-life "dirty" photos (as opposed to "clean" dataset photos) it's not there yet. **BUT** I do believe that's primarily due to not enough data. The methods and model architecture should be competitive with human facial recognition models and it should extend to other pets (there are just a lot more readily available human identity datasets out there). I've been talking with the Immich devs and it sounds like they were thinking we would probably need to crowd source our own dataset with user submitted images and labels.

All that to say: I'm trying to get the ball rolling so we can definitively outline the gaps to get it to production quality. It's probably a long way off from actual integration. I'm working on the models. Once the models are ready, we'll actually have to do the integration with the Immich UI and identity database. I started this 3 months ago, life got busy, I came back and worked on it some over the holidays. I will be working on this in my free time so don't get your hopes up too much. If I was working on this full time and hundreds of people were willing to submit their labeled pet data: maybe 2-3 months before an official beta. 

You might be wondering: **what's next**? There's one more public dataset that I need to get access to ([petface](https://dahlian00.github.io/PetFacePage/)) to feel like I've really gotten as much (currently available) public data as possible for animal identification. After that I will probably reach out to the Immich devs to see how they might want to go about crowd sourcing the dataset.

If you want to help, I'm open to contributions! BUT I'd prefer if you have ML experience. Though if you want to contribute to setting up the crowd sourcing process, then you don't really need ML experience 😁. The reason I say that is because I know the fundamentals for training these models and we're at the point of "how do we juice the most performance out of our data and models" which isn't particularly conducive to beginners and I don't have much time to provide guidance.

Thanks all for your interest in this feature. It's super important for me as well if I'm going to switch away from Google Photos. I'm time (and willpower) constrained, but hoping to make something out of this project that the open source community as a whole can benefit from.

**Reply by mFIND** [reply]
I was going to propose the same project! ([petface](https://dahlian00.github.io/PetFacePage/))
They suggest using [AnyFace](https://github.com/IS2AI/AnyFace) for face boundary, but seems that they (petface) only work with academia/research organizations.
Might be worth considering, if open source projects (like immich) would be provided access to animal data set, if they would be asked nicely.. :D

**Comment by mathmaniac43** [comment]
I am a relatively new Immich user, excited to see progress happening with this feature. Hope I can help train or test someday.

Early on, someone mentioned using the built-in search function to describe a pet and return matching images. I wonder if that could be used as a solution. For example, I have two dogs: one is all black, and one is mostly black with white snout, chest, feet, and tip of her tail (both are very cute, I digress). If I could make a map of search queries to associate with each pet, I bet it could do a pretty good job with sufficiently different-looking pets. For example, "black dog with white on their face, paws, and tip of tail" could map to Dog A, and "black dog with no white spots at all" could map to Dog B. Whenever images are uploaded, these queries could be run, just like the human facial recognition.

Assuming the existing search is good enough, this would mostly be a bit of UI work. And it could even work for less typical pets less likely to get dedicated models someday (e.g. guinea pigs), and non-pets too (e.g. "hot air balloons").

**Comment by ados8** [comment]
Apologies if this has already been mentioned, but I've used Nextcloud with Recognize which does many different animals. It wasn't always detecting correctly, but I found Viseron pet detection to be very good.

**Reply by seb-1204** [reply]
https://github.com/nextcloud/recognize

**Comment by mcattle** [comment]
It might also be a good idea to integrate the Merlin Bird ID model by Cambridge Labs if possible, so the species of pretty duck you photograph on a walk can be identified in Immich.  Would be great for birders.

Aside from that, pet bird detection by name would also be nice.

**Comment by scotthesterberg** [comment]
I have a working version of this request but it is not allowed be merged due to being created with AI. It is based on release 2.7.5, if anyone wants pet recognition it works for now, but it probably will not with new versions as the database changes will not align with changes to Immich in the future so use with great caution:

- https://github.com/scotthesterberg/immich/tree/feat/pet-recognition-final-fix
- https://github.com/scotthesterberg/immich-charts/tree/feat/pet-recognition-deployment
- https://github.com/scotthesterberg/ml-models/tree/feat/export-pet-models
- https://github.com/scotthesterberg/ui/tree/feat/pet-icons-components
- https://github.com/scotthesterberg/base-images/tree/feat/ml-pet-dependencies

**Reply by PseudoResonance** [reply]
Sorry, but I'm confused... From what I see, all this does is tells you that a pet is present. Did I misunderstand? I just don't see at all where this is grouping similar pets.

**Reply by scotthesterberg** [reply]
Honestly like I said I do not understand this tech well enough to explain it. I work lower in the stack, not as a developer. I can say that running containers built from this code (using Pet Recognition (MegaDescriptor - Non-Commercial): BVRA/MegaDescriptor-L-384, and Pet Detection (YOLO): Quazim0t0/yolov8-onnx) plugged in (did not build them in to avoid license complications, that's why I went with the pluggable models approach) it was able to recognize my dog and my mother in laws dog fairly accurately (seperately).

**Reply by scotthesterberg** [reply]
Oh I will say it failed to pick up my cat frustratingly, but since the models are somewhat pluggable in theory a better compatible model would fix that issue.

**Reply by scotthesterberg** [reply]
How to use Pet Recognition code above

  This guide covers building the custom containers and configuring the new "Bring Your Own Model" (BYOM) pet recognition system.

  1. Build the Custom Containers
  You need to build the updated versions of the main Immich server and the Machine Learning service from the source code.

    Build the Main Server:

     - podman build -t <your-repo>/immich-server:pet-configurable -f immich/server/Dockerfile immich

    Build the Machine Learning Service:

     - podman build -t <your-repo>/immich-machine-learning:pet-configurable -f immich/machine-learning/Dockerfile immich/machine-learning

  2. Update your Deployment
  Update your docker-compose.yml or Kubernetes values.yaml to use the new image tags.

  Provide a Hugging Face Token:
  To download the pet models, the Machine Learning container needs a Hugging Face "Read" token. Add it as an environment variable to the machine-learning container:

   `MACHINE_LEARNING_HF_TOKEN=your_hugging_face_token_here`

  3. Configure Pet Models in the Web UI
  Once the containers are running:
   1. Log in as an Administrator.
   2. Navigate to Administration -> Settings -> Machine Learning Settings.
   3. Scroll to the Pet Recognition section and enable the toggle.
   4. Enter the following Repository IDs:
       * Detection Model: Quazim0t0/yolov8-onnx
       * Recognition Model: s8648/MegaDescriptor-L-384 (or your own ONNX export)
   5. Click Save.

  4. Run the Recognition Jobs
  To process your existing library:
   1. Go to Administration -> Job Status.
   2. Click All under Pet Detection.
   3. Once detection finishes, click All under Pet Recognition.

  Your pets will now appear on the People page with a paw icon.

  ---

  Technical Note on the MegaDescriptor Model
  The original MegaDescriptor repository on Hugging Face (BVRA/MegaDescriptor-L-384) only contains PyTorch weights. Immich requires the ONNX format.

  If you want to use a different version of MegaDescriptor, you must use the export_megadescriptor.py script included in this repository to generate a .onnx file and upload it to your own Hugging Face account before providing that URL to the Web UI.

**Comment by tedornitier** [comment]
I built a tool for this: [immich-pet-tagger](https://github.com/tedornitier/immich-pet-tagger)

It runs as a Docker sidecar alongside Immich. The idea is that instead of needing a pretrained recognition model , you give it a few reference photos per pet through a small web UI, and it learns to recognize yours specifically using CLIP embeddings and YOLO. New photos get tagged automatically and your pets show up in the People section just like humans.

Works for cats, dogs, or anything visually distinct. Would love any feedback 🐾 

<img width="474" height="330" alt="screenshot" src="https://github.com/user-attachments/assets/e777167e-855c-45d3-a5c9-b8202e0f758d" />

**Reply by sunnysgithub** [reply]
Awesome!!!

**Reply by auanasgheps** [reply]
This looks AWESOME!

**Reply by mcattle** [reply]
Exactly what I needed!

**Reply by tianlang0704** [reply]
<img width="358" height="168" alt="image" src="https://github.com/user-attachments/assets/fa3d4d94-d094-4690-bad9-62c1833fac7f" />
Holy! I've been waiting for this for so long, and it even works for videos... Thank you so much! With a little bit of test and tweak, I feel this is even better than google photos pet detection.

**Reply by GiorgioMendieta** [reply]
This works really well, congratulations on implementing this!

**Reply by matoy** [reply]
waoh ; thank you so much, waiting for that since immich day 1

**Reply by gvolpe** [reply]
@tedornitier thanks a lot for sharing! I've had mixed results with it, can't fully automate it yet, so I've made "manual scanning mode" the default for me so I can review before a photo is tagged.

I've made quite a few changes in my [fork](https://github.com/gvolpe/immich-pet-tagger#fork-notice) tailored to my needs, specifically the ability to run it as a systemd service via the NixOS module. Still tweaking it but had good results from manual scans.

**Comment by shleeable** [comment]
I don't know if we can afford Jim Carrey.

<img width="686" height="386" alt="image" src="https://github.com/user-attachments/assets/c67a544e-429c-4c05-a464-dfcc8bf7d752" />

**Comment by rtp4jc** [comment]
Hi pet people! I have an update on my initial post on a pet detection model library. The models are ready for beta testing on dogs!

https://github.com/rtp4jc/immich-animals/releases/tag/sidecar-v0.1.0

I've gotten the models in a state where I can share it more broadly and it's good enough for me to start using them in my personal Immich library. Check it out and feel free to give me feedback or report issues via github.

<img width="1600" height="1000" alt="immich-people-dogs-named" src="https://github.com/user-attachments/assets/cd5b5106-474b-497d-b919-68bb97f9a82f" />

For some context, it runs as a docker sidecar similar to how @tedornitier 's solution works, but it uses state of the art models finetuned to dog detection and identification very similar to the existing person detection feature in Immich. It requires **no manual tweaking** to work out of the box aside from merging dog faces like you have to do for some humans. You're welcome to browse the main package for the current benchmarking data on detection and identity embeddings, but you should be able to just boot up the sidecar without having to worry about that.

**What's next?** The models mostly need more data to improve and expand to new species. I still have not gotten access to the petface dataset and may not be able to as the maintainers stopped responding a while back. I may be able to recreate what petface did manually at a smaller scale (all of the photos in that dataset are from the internet), but a crowd-sourced dataset is an alternative. I plan to work with the Immich team as I know they had plans for this kind of crowd-sourced dataset prep as future work and I could help out there.

**Reply by Gregor1996** [reply]
I want to highlight your request for a crowd-sourced pets dataset!

If you, or the Immich team himself, provides a way to upload the correctly tagged photos I could start helping with this!
